narwhals/typing.py

```diff
@@ -37,6 +45,48 @@ class DataFrameLike(Protocol):
         def __dataframe__(self, *args: Any, **kwargs: Any) -> Any: ...
 
 
+class CompliantSeries(Protocol):
+    @property
+    def name(self) -> str: ...
+    def __narwhals_series__(self) -> CompliantSeries: ...
+    def alias(self, name: str) -> Self: ...
+
+
+class CompliantDataFrame(Protocol):
+    def __narwhals_dataframe__(self) -> CompliantDataFrame: ...
+    def __narwhals_namespace__(self) -> Any: ...
+
+
+class CompliantLazyFrame(Protocol):
+    def __narwhals_lazyframe__(self) -> CompliantLazyFrame: ...
+    def __narwhals_namespace__(self) -> Any: ...
+
+
+CompliantSeriesT_co = TypeVar(
+    "CompliantSeriesT_co", bound=CompliantSeries, covariant=True
+)
+
+
+@runtime_checkable
+class CompliantExpr(Protocol, Generic[CompliantSeriesT_co]):
+    _implementation: Implementation
+    _output_names: list[str] | None
+    _root_names: list[str] | None
+    _depth: int
+    _function_name: str
```

Comment:
@MarcoGorelli if possible, I'd recommend using the `typing_extensions` versions of `Protocol`, `@runtime_checkable`

- https://github.com/python/cpython/pull/103034

It has quite a lot of backports for performance/bugfixes:

- https://github.com/search?q=repo%3Apython%2Ftyping_extensions%20%22protocol%22%20path%3A%22CHANGELOG.md%22&type=code

Reply:
Also, you *might* be able to do this with `typing` in `3.9`, but **definitely** with `typing_extensions`:

```py
@runtime_checkable
class CompliantExpr(Protocol[CompliantSeriesT_co]):...
```

Reply:
thanks @dangotbanned ! you're really the #1 typing expert

I'd like to keep things zero-dependency in Narwhals, so I think I'll just check for the `__narwhals_` dunders

Reply:
> thanks @dangotbanned ! you're really the https://github.com/narwhals-dev/narwhals/issues/1 typing expert

😄 

> I'd like to keep things zero-dependency in Narwhals, so I think I'll just check for the __narwhals_ dunders

Totally understandable!